viz/TraversabilityGridVisualization.cpp

```diff
@@ -0,0 +1,325 @@
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <vizkit3d/TraversabilityGridVisualization.hpp>
+#include <vizkit3d/ExtentsRectangle.hpp>
+#include <vizkit3d/ColorConversionHelper.hpp>
+
+#include <osg/Texture2D>
+
+
+using namespace vizkit3d;
+
+TraversabilityGridVisualization::TraversabilityGridVisualization()
+    : showMapExtents(true),
+      interpolateCellColors(true),
+      useNPOTTextures(false),
+      unknownColor(osg::Vec4(0, 0, 1, 1)),
+      obstacleColor(osg::Vec4(0.8, 0, 0, 1)),
+      minTraversableColor(osg::Vec4(1, 0, 0, 1)),
+      maxTraversableColor(osg::Vec4(0, 1, 0, 1))
+{
+}
+
+TraversabilityGridVisualization::~TraversabilityGridVisualization()
+{
+}
+
+osg::ref_ptr<osg::Node> TraversabilityGridVisualization::createMainNode()
+{
+    // Geode is a common node used for vizkit3d plugins. It allows to display
+    // "arbitrary" geometries
+    osg::ref_ptr<osg::Group> group = new osg::Group();
+
+    planeGeode = new osg::Geode();
+    plane = createPlane();
+    planeGeode->addDrawable(plane.get());
+    group->addChild(planeGeode.get());
+
+    extentsRectangle = new ExtentsRectangle(Eigen::Vector2d(), Eigen::Vector2d());
+
+    return group;
+}
+
+void TraversabilityGridVisualization::updateMainNode(osg::Node* node)
+{
+    osg::ref_ptr<osg::Group> group = node->asGroup();
+
+    visualize(*planeGeode);
+
+    visualizeMapExtents(*group);
+}
+
+void TraversabilityGridVisualization::updateDataIntern(const maps::grid::TraversabilityGrid& plan)
+{
+    traversabilityGrid = plan;
+}
+
+void vizkit3d::TraversabilityGridVisualization::visualize(osg::Geode& geode) const
+{
+        updatePlane();
+
+        osg::ref_ptr<osg::StateSet> state = geode.getOrCreateStateSet();
+
+        osg::ref_ptr<osg::Image> image = createTextureImage();
+
+        osg::ref_ptr<osg::Texture2D> tex = new osg::Texture2D(image.get());
+        tex->setResizeNonPowerOfTwoHint(!useNPOTTextures);
+
+        if (!interpolateCellColors)
+        {
+            tex->setFilter(osg::Texture::MIN_FILTER, osg::Texture::NEAREST);
+            tex->setFilter(osg::Texture::MAG_FILTER, osg::Texture::NEAREST);
+        }
+
+        state->setTextureAttributeAndModes(0, tex.get());
+}
+
+void TraversabilityGridVisualization::visualizeMapExtents(osg::Group& group) const
+{
+    unsigned int numChildren = group.getNumChildren();
+
+    if (showMapExtents)
+    {
+        if (numChildren <= 1)
```

Comment:
Please note that you can ask group if it has a specific child. See [containsNode](http://camlosg.sourceforge.net/osg/classosg_1_1Group.html#5b7d6aa9a106792464684790f6b377e4).